freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserSyncService.java

```diff
@@ -386,42 +398,60 @@ private boolean usersStateDifferenceChanged(UsersStateDifference beforeSync, Use
 
         LOGGER.debug("Starting {} ...", LogEvent.CALCULATE_UMS_IPA_DIFFERENCE);
-        UsersStateDifference usersStateDifference = UsersStateDifference.fromUmsAndIpaUsersStates(umsUsersState, ipaUsersState);
+        UsersStateDifference usersStateDifference = UsersStateDifference.fromUmsAndIpaUsersStates(umsUsersState, ipaUsersState,
+                credentialsUpdateOptimizationEnabled);
         LOGGER.debug("Finished {}.", LogEvent.CALCULATE_UMS_IPA_DIFFERENCE);
 
         return usersStateDifference;
     }
 
     private void applyDifference(UmsUsersState umsUsersState, String environmentCrn, Multimap<String, String> warnings,
-            UsersStateDifference usersStateDifference, FreeIpaClient freeIpaClient, boolean fmsToFreeipaBatchCallEnabled)
-            throws FreeIpaClientException, IOException {
+            UsersStateDifference usersStateDifference, FreeIpaClient freeIpaClient, boolean fmsToFreeipaBatchCallEnabled,
+            boolean credentialsUpdateOptimizationEnabled) throws FreeIpaClientException, IOException {
         LOGGER.debug("Starting {} ...", LogEvent.APPLY_DIFFERENCE_TO_IPA);
         applyStateDifferenceToIpa(environmentCrn, freeIpaClient, usersStateDifference, warnings::put, fmsToFreeipaBatchCallEnabled);
         LOGGER.debug("Finished {}.", LogEvent.APPLY_DIFFERENCE_TO_IPA);
 
         if (!FreeIpaCapabilities.hasSetPasswordHashSupport(freeIpaClient.getConfig())) {
             LOGGER.debug("IPA doesn't have password hash support, no credentials sync required for env:{}", environmentCrn);
         } else {
-            // Sync credentials for all users and not just diff. At present there is no way to identify that there is a change in password for a user
-            LOGGER.debug("Starting {} for {} users ...", LogEvent.SET_WORKLOAD_CREDENTIALS, umsUsersState.getUsersWorkloadCredentialMap().size());
-            workloadCredentialService.setWorkloadCredentials(fmsToFreeipaBatchCallEnabled, freeIpaClient,
-                    umsUsersState.getUsersWorkloadCredentialMap(), warnings::put);
+            // If the credentials update optimization is enabled, sync credentials for users with updated credentials.
+            // Otherwise, sync credentials for all users.
+            LOGGER.debug("Starting {} for {} users ...", LogEvent.SET_WORKLOAD_CREDENTIALS, usersStateDifference.getUsersWithCredentialsToUpdate().size());
+            Map<String, Pair<String, WorkloadCredential>> userToCredentialUpdateParamsMap =
+                    usersStateDifference.getUsersWithCredentialsToUpdate().stream().collect(Collectors.toMap(
+                            Function.identity(), username -> getCredentialUpdateParams(username, umsUsersState)));
+            workloadCredentialService.setWorkloadCredentials(fmsToFreeipaBatchCallEnabled, credentialsUpdateOptimizationEnabled, freeIpaClient,
+                    userToCredentialUpdateParamsMap, warnings::put);
```

Comment:
these boolean stuff getting a bit out of hand, and if you need to comment makes me think something is fishy here.
So your comment would suggest that at this point we will branch the logic depending if this feature is enabled and sync all users or only the updated one. But this is not the case. All users will be updated which are in `usersStateDifference.getUsersWithCredentialsToUpdate()`
I think you should simply drop this comment.

Reply:
Comment dropped.